pkg/s3select/json/reader.go

```diff
@@ -59,124 +60,20 @@ func toSingleLineJSON(input string, currentKey string, result gjson.Result) (out
 	return output, err
 }
 
-type objectReader struct {
-	reader io.Reader
-	err    error
-
-	p     []byte
-	start int
-	end   int
-
-	escaped     bool
-	quoteOpened bool
-	curlyCount  uint64
-	endOfObject bool
-}
-
-func (or *objectReader) objectEndIndex(p []byte, length int) int {
-	for i := 0; i < length; i++ {
-		if p[i] == '\\' {
-			or.escaped = !or.escaped
-			continue
-		}
-
-		if p[i] == '"' && !or.escaped {
-			or.quoteOpened = !or.quoteOpened
-		}
-
-		or.escaped = false
-
-		switch p[i] {
-		case '{':
-			if !or.quoteOpened {
-				or.curlyCount++
-			}
-		case '}':
-			if or.quoteOpened || or.curlyCount == 0 {
-				break
-			}
-
-			if or.curlyCount--; or.curlyCount == 0 {
-				return i + 1
-			}
-		}
-	}
-
-	return -1
-}
-
-func (or *objectReader) Read(p []byte) (n int, err error) {
-	if or.endOfObject {
-		return 0, io.EOF
-	}
-
-	if or.p != nil {
-		n = copy(p, or.p[or.start:or.end])
-		or.start += n
-		if or.start == or.end {
-			// made full copy.
-			or.p = nil
```

Comment:
The actual problem is it forgets saved `or.p` here.  The fix is below
```diff
diff --git a/pkg/s3select/json/reader.go b/pkg/s3select/json/reader.go
index fb4e15ba..3bfc7b57 100644
--- a/pkg/s3select/json/reader.go
+++ b/pkg/s3select/json/reader.go
@@ -112,13 +112,6 @@ func (or *objectReader) Read(p []byte) (n int, err error) {
 
 	if or.p != nil {
 		n = copy(p, or.p[or.start:or.end])
-		or.start += n
-		if or.start == or.end {
-			// made full copy.
-			or.p = nil
-			or.start = 0
-			or.end = 0
-		}
 	} else {
 		if or.err != nil {
 			return 0, or.err
@@ -137,6 +130,16 @@ func (or *objectReader) Read(p []byte) (n int, err error) {
 
 	index := or.objectEndIndex(p, n)
 	if index == -1 || index == n {
+		if or.p != nil {
+			or.start += n
+			if or.start == or.end {
+				// made full copy.
+				or.p = nil
+				or.start = 0
+				or.end = 0
+			}
+		}
+
 		return n, nil
 	}
 
@@ -146,7 +149,7 @@ func (or *objectReader) Read(p []byte) (n int, err error) {
 		or.start = index
 		or.end = n
 	} else {
-		or.start -= index
+		or.start += index
 	}
 
 	return index, nil
```

It works for
```bash
$ for i in $(seq 10 10 100); do head -$i bug.json | mc pipe local/mybucket/bug-$i.json ; done
$ for i in $(seq 10 10 100); do  mc sql --query 'SELECT * FROM S3Object' local/mybucket/bug-$i.json | wc -l ; done
10
20
30
40
50
60
70
80
90
100
```

Similarly, test code
```go
func TestObjectReader(t *testing.T) {
	f, err := os.Open("bug.json")
	if err != nil {
		panic(err)
	}

	or := &objectReader{reader: f}

	for {
		if err := or.Reset(); err != nil {
			if err == io.EOF {
				break
			}

			panic(err)
		}

		data, err := ioutil.ReadAll(or)
		if err != nil {
			panic(err)
		}

		data = bytes.TrimSpace(data)
		if len(data) == 0 {
			break
		}

		if data[0] != '{' {
			panic("broken json")
		}

		fmt.Println(string(data))
	}
}
```

returns data is exactly same as the source.
```bash
$ md5sum /tmp/genbug.json bug.json 
e3a380d997fd861cadca02a142094aef  /tmp/genbug.json
e3a380d997fd861cadca02a142094aef  bug.json
```

Reply:
I really don't want to patch that implementation, it is too cryptic and complex. This fix is going to make it more complex.

I prefer that the new approach be reviewed, it is simpler and uses Go standard libraries to provide similar behavior.

Reply:
PR fails for below inputs
1. ```json
{"a":"}"
}
```
2. ```json
{
   "foo": {
       "bar": "baz"
   }
}
```

You would need to use similar logic in lexer `objectReader.objectEndIndex()`. 

When memory is concern, `scanJSON()` is called with appended data of every iteration.  For larger input, it is called with multiples of 4K, e.g. 4K, 8K, 16K and so on.  On every iteration, the data is available from the beginning i.e. `scanJSON()` scans from index 0 to len(data).  Similarly tokenizing involves fresh memory allocation.

comparing this with `objectReader`, `objectReader` is little memory efficient where no memory allocation involved.  This may be advantageous for huge JSON files.